package/MDAnalysis/analysis/base.py

```diff
@@ -430,25 +517,219 @@ def run(self, start=None, stop=None, step=None, frames=None,
         verbose = getattr(self, '_verbose',
                           False) if verbose is None else verbose
 
-        self._setup_frames(self._trajectory, start=start, stop=stop,
-                           step=step, frames=frames)
         logger.info("Starting preparation")
-        self._prepare()
         logger.info("Starting analysis loop over %d trajectory frames",
                     self.n_frames)
 
-        for i, ts in enumerate(ProgressBar(
-                self._sliced_trajectory,
+        if len(indexed_frames) == 0: # if `frames` were empty in `run` or `stop=0`
+            return self
+        
+        frame_indices, frames = indexed_frames[:, 0], indexed_frames[:, 1], 
+        trajectory = self._trajectory[frames]
+        for idx, ts in enumerate(ProgressBar(trajectory,
                 verbose=verbose,
                 **progressbar_kwargs)):
+            i = frame_indices[idx]
             self._frame_index = i
             self._ts = ts
             self.frames[i] = ts.frame
             self.times[i] = ts.time
             self._single_frame()
         logger.info("Finishing up")
+        return self
+
+    def _setup_computation_groups(self, n_parts: int, start=None, stop=None, step=None, frames=None) -> list[np.ndarray]: 
+        """
+        Set the self._computation_groups (the workload distribution scheme) for the future delayed
+        computations.
+
```

Comment:
This function is very important, I'd add more details here in the docstring. Just saying that `Set the self._computation_groups` is OK for the beginning of the docstring, but it is not explaining much.

Reply:
I think it's better now